### PR TITLE
feat(mdlint): add rule TOP009 to ensure lesson overview items start with a capital letter and end with a period

### PR DESCRIPTION
**Pull request by nikitarevenco**
Adds a new rule to markdown lint.

Lesson overview items should end with a period and start with a capital letter

## Issue

Closes https://github.com/TheOdinProject/curriculum/issues/27625#issuecomment-2112275087

## Pull Request Requirements

-   [x] I have thoroughly read and understand [The Odin Project curriculum contributing guide](https://github.com/TheOdinProject/curriculum/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`
-   [x] The `Because` section summarizes the reason for this PR
-   [x] The `This PR` section has a bullet point list describing the changes in this PR
-   [x] If this PR addresses an open issue, it is linked in the `Issue` section
-   [x] If any lesson files are included in this PR, they have been previewed with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure it is formatted correctly
-   [x] If any lesson files are included in this PR, they follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

### CODE REVIEW

**Review by MaoShizhong — CHANGES_REQUESTED** [review]
The checks aren't running due to a merge conflict.
You'll need to rename your rule and relevant files/dirs to use `TOP009` (including the markdownlint config) because #27915 added what's now `TOP008`. That PR also adjusted the markdownlint config, so that should be resolved once you resolve the merge conflicts.

In terms of testing your rule, you'll need to make sure your test files include all of the different scenarios to test, and you can run the lint script on it locally.
Once you're happy with that, then a new push to here (after resolving conflicts) will trigger the actions with your local linting files.

**Comment by nikitarevenco** [comment]
> The checks aren't running due to a merge conflict. You'll need to rename your rule and relevant files/dirs to use `TOP009` (including the markdownlint config) because #27915 added what's now `TOP008`. That PR also adjusted the markdownlint config, so that should be resolved once you resolve the merge conflicts.
> 
> In terms of testing your rule, you'll need to make sure your test files include all of the different scenarios to test, and you can run the lint script on it locally. Once you're happy with that, then a new push to here (after resolving conflicts) will trigger the actions with your local linting files.

Thanks for letting me know! I've fixed the merge conflicts. Now I just need to fix the code so that it works

**Comment by MaoShizhong** [comment]
A heads up that the merge has wiped out your original doc file, so you'll want to restore it but under the TOP009 name of course

**Comment by nikitarevenco** [comment]
> A heads up that the merge has wiped out your original doc file, so you'll want to restore it but under the TOP009 name of course

Yay! I got it working

**Comment by MaoShizhong** [comment]
Question - my understanding from the original issue was that this rule was meant to **prevent** LO items being questions, as they instead should be statements of what the lesson will cover. But this rule tries to enforce the opposite - is this intended?

**Comment by nikitarevenco** [comment]
> Question - my understanding from the original issue was that this rule was meant to **prevent** LO items being questions, as they instead should be statements of what the lesson will cover. But this rule tries to enforce the opposite - is this intended?

Oh, this was just a mistake on my part. I just fixed it!

**Comment by nikitarevenco** [comment]
Actually, seems like there are still some mistakes. Sorry let me test it a bit more.

**Comment by nikitarevenco** [comment]
> Question - my understanding from the original issue was that this rule was meant to **prevent** LO items being questions, as they instead should be statements of what the lesson will cover. But this rule tries to enforce the opposite - is this intended? I see that the doc file implies the rule is supposed to prevent questions (LO items must begin capitalised and end with a `.`)

The tests are passing as expected now. I just accidentally searched for the 2nd character instead of the 1st character. `.at(1)`. I have been writing code in another language where the array indexes start with `1` so I didn't think twice about this

File: markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  return {
+    firstCharacter,
+    lastCharacter,
+    lineNumber,
+    context,
+  };
+}
+
+module.exports = {
+  names: ["TOP009", "lessonOverviewItemsSentenceStructure"],
```
Comment (MaoShizhong) [inline markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js:40]:
```suggestion
  names: ["TOP009", "lesson-overview-items-sentence-structure"],
```
Kebab case for consistency between other rules (the lint report shows the rule name by doing `.names.join("/")`

File: markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      const firstCharacterIsValid =
+        firstCharacter === firstCharacter.toUpperCase();
+      if (!firstCharacterIsValid) {
+        onError({
+          lineNumber,
+          detail: "Lesson overview items should start with a capital letter",
+          context,
+          fixInfo: {
+            lineNumber,
+            editColumn: 2,
```
Comment (MaoShizhong) [inline markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js:60]: You're going to want to check `editColumn` for the fixer

File: markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      }
+      const lastCharacterIsValid = lastCharacter === ".";
+      if (!lastCharacterIsValid) {
+        onError({
+          lineNumber,
+          detail: "Lesson overview items should end with a period.",
+          context,
+          fixInfo: {
+            lineNumber,
+            editColumn: context.length,
```
Comment (MaoShizhong) [inline markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js:74]: You'll also want to check the editColumn here as well

File: markdownlint/docs/TOP009.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# `TOP009` - Lesson overview items sentence structure
+
+Tags: `content`, `language`
```
Comment (MaoShizhong) [inline markdownlint/docs/TOP009.md:3]:
```suggestion
Tags: `content`
```
Since `content` is the only tag in the rule (and `language` is more for the languages used like code block language)

File: markdownlint/TOP009_lessonOverviewItemsSentenceStructure/tests/TOP009_no_punctuation.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
+### Introduction
+
+This file should be marked with an error because some of the lesson overview items has no punctuation
```
Comment (MaoShizhong) [inline markdownlint/TOP009_lessonOverviewItemsSentenceStructure/tests/TOP009_no_punctuation.md:3]:
```suggestion
This file should be marked with an error because some of the lesson overview items do not end with a period.
```

File: markdownlint/TOP009_lessonOverviewItemsSentenceStructure/tests/TOP009_capital_letter.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
+### Introduction
+
+This file should be marked with an error because one of the lesson overview items is not properly capitalised.
```
Comment (MaoShizhong) [inline markdownlint/TOP009_lessonOverviewItemsSentenceStructure/tests/TOP009_capital_letter.md:3]:
```suggestion
This file should be marked with an error because one of the lesson overview items is not properly capitalized.
```
Unfortunately, we Brits must conform

File: markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+            deleteCount: 1,
```
Comment (MaoShizhong) [inline markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js:75]: While a delete count of `1` might suffice for most realistic situations, what about in the instance of 
```text
- Lesson overview item?.
- Lesson overview item???.
- Lesson overview item...
```
Don't necessarily have to overcomplicate for *every* possible edge case but it probably wouldn't be that much more complex to handle just chopping off all chars past the last letter, instead of hard coding a deleteCount of 1.

**Review by MaoShizhong — CHANGES_REQUESTED** [review]
Great work @nikitarevenco :rocket:

A few nits, a few things to consider and a few little bugs to address, but otherwise, the bulk of this is :fire:

@thatblindgeye - would also appreciate your run through of this whenever you get the chance, in case you had any further insights or ideas.

File: markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+module.exports = {
+  names: ["TOP009", "lessonOverviewItemsSentenceStructure"],
+  description: "Lesson overview items should have proper punctuation",
+  tags: ["content"],
+  information: new URL(
+    "https://github.com/TheOdinProject/curriculum/blob/main/markdownlint/docs/TOP009.md",
+  ),
+  function: function TOP008(params, onError) {
```
Comment (thatblindgeye) [inline markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js:46]:
```suggestion
  function: function TOP009(params, onError) {
```

File: markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    const bulletPoints = isolateBulletList(params.tokens);
+    bulletPoints.forEach((bulletPoint) => {
+      const { firstCharacter, lastCharacter, lineNumber, context } =
+        getListItemData(bulletPoint);
+      const firstCharacterIsValid =
+        firstCharacter === firstCharacter.toUpperCase();
+      if (!firstCharacterIsValid) {
+        onError({
+          lineNumber,
+          detail: "Lesson overview items should start with a capital letter",
```
Comment (thatblindgeye) [inline markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js:56]:
```suggestion
          detail: "Lesson overview items must start with a capital letter",
```
Can't think of a case where we wouldn't have a capital letter to start, so "must" will convey that it is necessary.

File: markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+            deleteCount: 1,
+            insertText: firstCharacter.toUpperCase(),
+          },
+        });
+      }
+      const lastCharacterIsValid = lastCharacter === ".";
+      if (!lastCharacterIsValid) {
+        onError({
+          lineNumber,
+          detail: "Lesson overview items should end with a period.",
```
Comment (thatblindgeye) [inline markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js:70]:
```suggestion
          detail: "Lesson overview items must end with a period.",
```
Though also maybe we could check whether the item ends with a question mark, and make the error message, "Lesson overview items should end with a period and be written as a statement rather than a question." That isn't a blocker, though, so I'm fine with just the single message.

File: markdownlint/TOP009_lessonOverviewItemsSentenceStructure/tests/TOP009_invalid_punctuation.md [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+This file should be marked with an error because one of the lesson overview items is a question.
+
+### Lesson overview
+
+This section contains a general overview of topics that you will learn in this lesson.
+
+- Lesson overview item 1.
+- Lesson overview item 2.
+- Lesson overview item 3?
```
Comment (thatblindgeye) [inline markdownlint/TOP009_lessonOverviewItemsSentenceStructure/tests/TOP009_invalid_punctuation.md:11]: I feel like we could combine this and the "no_punctuation" file into one test, since they're essentially testing for the same thing (valid punctuation).

File: markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js [hunk 12]
```diff
@@ -0,0 +1,82 @@
+// A bullet list is every token between and inclusive bullet_list_open to bullet_list_closed
```
Comment (thatblindgeye) [inline markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js:1]: I don't think we really need these comments in the file

**Review by thatblindgeye — CHANGES_REQUESTED** [review]
Awesome job on this 🔥

File: markdownlint/docs/TOP009.md [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+```markdown
+- Lesson overview item 1.
+- Lesson overview item 2.
+- Lesson overview item 3.
+```
+
+## Rationale
+
+Make sure lesson overview items follow a consistent structure. Lesson overview items should be brief overviews of what the lesson is about, they should not be questions.
```
[the remark this answers is not in the document]
Comment (MaoShizhong) [reply markdownlint/docs/TOP009.md:25]: Think we could chop off the "that the user is expected to be able to answer after completing the lesson" bit, so just the following reads :ok_hand: to me.

> Lesson overview items should follow a consistent structure. They must tell the user what the lesson is about via brief overview statements, and must not be conveyed in the form of a question.

File: markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js (hunk 7 above)
Comment (MaoShizhong) [reply markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js:75]: Testing for `\w\.$` is simple enough, though in terms of the fixer, it'd have to count the trailing punctuation to provide the correct `fixInfo.deleteCount` number. Just tested and providing a higher number than there are chars in the rest of the line unfortunately throws an error. No quick and dirty `deleteCount: Infinity` sadly!

Comment (MaoShizhong) [reply markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js:75]: @thatblindgeye What do you think about links in LO items? I notice a few lessons, particularly in the Rails pathway, have links in LO items. This just came to mind for an edge case:
```markdown
- Blah blah [some link text](href).
```
Since if we did test for the char preceding the first of the trailing punctuation, we'd have to include `)` and also amend the rule+docs to account for (or reject) links in LO items.

**Comment by nikitarevenco** [comment]
For the changes that would be better made on my computer than via the github online editor, I'll add them in a few weeks time since I have exams at the moment.

File: markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js (hunk 10 above)
Comment (nikitarevenco) [reply markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js:72]: I think that's a good idea!

File: markdownlint/docs/TOP009.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (nikitarevenco) [reply markdownlint/docs/TOP009.md:25]: Yeah, that's better 100%

File: markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js (hunk 12 above)
Comment (nikitarevenco) [reply markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js:1]: Oh yeah, forgot to remove them - They're misleading

**Comment by MaoShizhong** [comment]
> For the changes that would be better made on my computer than via the github online editor, I'll add them in a few weeks time since I have exams at the moment.

Not a problem at all - there's no rush for this rule for sure, and even if there somehow was, your exams are more important so please do focus on those first and foremost!
Whatever changes are needed for this PR, you can certainly sort them out when you're done and have the time for this, since this is all voluntary.
We have a TOP010 rule in review as well, but we can always merge that first if appropriate then just resolve the merge conflict in the custom rules array when you get back to this later.

**Comment by nikitarevenco** [comment]
Not sure what's going on here, I've removed the comment but that didn't seem to work

image

used to work just fine before

File: .markdownlint-cli2.jsonc [hunk 14]
```diff
@@ -87,6 +86,7 @@
     "./markdownlint/TOP006_fullFencedCodeLanguage/TOP006_fullFencedCodeLanguage.js",
     "./markdownlint/TOP007_useMarkdownLinks/TOP007_useMarkdownLinks.js",
     "./markdownlint/TOP008_useBackticksForFencedCodeBlocks/TOP008_useBackticksForFencedCodeBlocks.js",
-    "./markdownlint/TOP010_useLazyNumbering/TOP010_useLazyNumbering.js"
+    "./markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js",
+    "./markdownlint/TOP010_useLazyNumbering/TOP010_useLazyNumbering.js",
```
Comment (MaoShizhong) [inline .markdownlint-cli2.jsonc:90]: JSON doesn't support trailing commas, hence the syntax error being raised
```suggestion
    "./markdownlint/TOP010_useLazyNumbering/TOP010_useLazyNumbering.js"
```

**Comment by nikitarevenco** [comment]
@MaoShizhong  Thank you! The PR is not ready for review yet as the tests are not passing. I will let you know when they are.

**Review by MaoShizhong — COMMENTED** [review]
Had a think about this further.

I think we should add a non-script-fixable error if an LO item ends with a `?` that reports something like "Lesson overview items must be statements, not questions."

If I wrote `- What are linters?` then the current code would only complain about the `?` and simply fixing (script or manual) to `- What are linters.` will not make sense, yet the linter will be happy.

I think questions should be flagged as such and *not* made script-fixable, as that requires changing the sentence wording itself e.g. to `- Describe what linters are.`

Then in that situation, we can flag and script-fix capitalisation/end punctuation errors.

**Comment by nikitarevenco** [comment]
> Had a think about this further.
> 
> I think we should add a non-script-fixable error if an LO item ends with a `?` that reports something like "Lesson overview items must be statements, not questions."
> 
> If I wrote `- What are linters?` then the current code would only complain about the `?` and simply fixing (script or manual) to `- What are linters.` will not make sense, yet the linter will be happy.
> 
> I think questions should be flagged as such and _not_ made script-fixable, as that requires changing the sentence wording itself e.g. to `- Describe what linters are.`
> 
> Then in that situation, we can flag and script-fix capitalisation/end punctuation errors. Those errors must only flag if the question error doesn't.

I agree with you. I think introducing fixers only makes sense when they are beneficial in every case.
I don't think it makes sense for there to be a punctuation auto fixer in the first place. For example, while `Describe what linters are` -> `Describe what linters are.` makes sense, `What are linters` -> `What are linters.` would not make as much sense because the sentence itself is phrased as a question, what do you think

**Comment by MaoShizhong** [comment]
I think I'm with you on that, where the nuances and complexity of a fixer catching all cases isn't really that beneficial given the example you provided.

I would probably think then we just need the one general error reporting something like
"Lesson overview items must be statements, not questions, and must begin with a capital letter and end with a period."
No need for actual/expected, just provide the LO item itself in the `context` property.

That should flag for questions (ends in ?), as well as uncapitalised/un-period-ended LO items. Then we leave it to the user to manually fix, since it's context-dependent.

**Comment by nikitarevenco** [comment]
> I think I'm with you on that, where the nuances and complexity of a fixer catching all cases isn't really that beneficial given the example you provided.
> 
> I would probably think then we just need the one general error reporting something like "Lesson overview items must be statements, not questions, and must begin with a capital letter and end with a period." No need for actual/expected, just provide the LO item itself in the `context` property.
> 
> That should flag for questions (ends in ?), as well as uncapitalised/un-period-ended LO items. Then we leave it to the user to manually fix, since it's context-dependent.

I made the lint rules behave like your suggestions:
- Single warning message `Lesson overview items must be statements, not questions, and must begin with a capital letter and end with a period.`

File: markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+        getListItemData(bulletPoint);
+      const firstCharacterIsValid =
+        firstCharacter === firstCharacter.toUpperCase();
+      const lastCharacterIsValid = lastCharacter === ".";
+      if (!firstCharacterIsValid || !lastCharacterIsValid) {
+        onError({
+          lineNumber,
+          detail:
+            "Lesson overview items must be statements, not questions, and must begin with a capital letter and end with a period.",
+          context,
```
Comment (MaoShizhong) [inline markdownlint/TOP009_lessonOverviewItemsSentenceStructure/TOP009_lessonOverviewItemsSentenceStructure.js:61]: Nit for consistency, both with how other files have their objects, and also avoiding mixing shorthand in an object with non-shorthand.
```suggestion
          lineNumber: lineNumber,
          detail:
            "Lesson overview items must be statements, not questions, and must begin with a capital letter and end with a period.",
          context: context,
```

File: markdownlint/docs/TOP008.md [hunk 16]
```diff
@@ -20,4 +20,4 @@ You should use backticks instead
 
 Consistent formatting makes it easier to understand a document.
 
-Markdown lint's [MD048 rule](https://github.com/DavidAnson/markdownlint/blob/main/doc/md048.md) already covers this check, but does not include fix information, therefore can only be used to raise errors for manual fixing. This custom rule enforces the same style but includes fix information that can be used alongside our fix scripts.
+Markdown lint's [MD048 rule](https://github.com/DavidAnson/markdownlint/blob/main/doc/md048.md) already covers this check, but does not include fix information, therefore can only be used to raise errors for manual fixing. This custom rule enforces the same style but includes fix information that can be used alongside our fix scripts.
```
Comment (MaoShizhong) [inline markdownlint/docs/TOP008.md:23]: No idea how this made it into the diff, given it's not showing any actual changes... Shouldn't need to do anything with it :shrug:

File: markdownlint/docs/TOP009.md [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+
+This rule is triggered when a lesson overview item doesn't end with a period `.` or doesn't start with a capital letter. For example:
+
+```markdown
+- Lesson overview item 1
+- Lesson overview item 2?
+- lesson overview item 3.
+```
+
+To fix this issue, make sure that each bullet point ends with a period `.` and starts with a capital letter.
```
Comment (MaoShizhong) [inline markdownlint/docs/TOP009.md:15]: Nit, even if the rational section does expand on it later.
```suggestion
To fix this issue, make sure that each bullet point ends with a period `.` and starts with a capital letter, rewording if you need to fix grammar since these should not be questions.
```

**Review by MaoShizhong — COMMENTED** [review]
Just two little nits but otherwise I'm happy with this :fire: Seems to work great!

File: markdownlint/docs/TOP008.md (hunk 16 above)
Comment (nikitarevenco) [reply markdownlint/docs/TOP008.md:23]: Yeah I was confused sitting there trying to figure out why this file is in here lol

File: markdownlint/docs/TOP009.md [hunk 18]
```diff
@@ -0,0 +1,25 @@
+# `TOP009` - Lesson overview items sentence structure
```
Comment (MaoShizhong) [inline markdownlint/docs/TOP009.md:1]: One more from me... in preparation for #28330

```suggestion
# `TOP009`: Lesson overview items sentence structure
```

**Review by MaoShizhong — APPROVED** [review]
I'm happy with this now - let's see if @thatblindgeye has anything further to add.

**Review by MaoShizhong — APPROVED** [review]
:pray:
